### PR TITLE
Subject groups creation bug fixes

### PR DESCRIPTION
**Pull request by camallen**
Linked to #3549 and #3550  add more checks and validation when creating the special `SubjectGroup.group_subject` Subject resource

Adds validations that the
- group we are creating matches our expected number of subjects
- group is bigger than 1 (else why are we creating a group?)

These will help with issues reported in the above issues where we are incorrectly creating groups

Finally there is a big bug fix in here that ensures we correctly create the `group_subjects` linked image media URLs in the order that they are supplied i.e. the subject ids match the linked images and they aren't mixed. Without this check we may actually show the incorrect image to the user and record the wrong subject ID in the downstream classification - which would corrupt the data and be a very bad thing. 

FWIW - i've added some data tracking metadata on the group subject media locations to record the originating subject resource ids to ensure we can reconstruct the correct ordering / linkage and not corrupt downstream classification data.

### TODO 
- [ ] Ensure each originating subject only has 1 media location as outlined in https://github.com/zooniverse/panoptes/issues/3549#issuecomment-760258970

# Review checklist

- [ ] First, the most important one: is this PR small enough that you can actually review it? Feel free to just reject a branch if the changes are hard to review due to the length of the diff.
- [ ] If there are any migrations, will they the previous version of the app work correctly after they've been run (e.g. the don't remove columns still known about by ActiveRecord).
- [ ] If anything changed with regards to the public API, are those changes also documented in the `apiary.apib` file?
- [ ] Are all the changes covered by tests? Think about any possible edge cases that might be left untested.

### CODE REVIEW

File: app/operations/subject_groups/create.rb [hunk 1]
```diff
@@ -83,12 +96,27 @@ def build_group_subject
...
         extension = File.extname(loc.src).downcase[1..-1]
         mime_type = Mime::Type.lookup_by_extension(extension).to_s
-        { mime_type => "https://#{loc.src}" }
+        media_url = "https://#{loc.src}"
+        # track the subject URL and the originating subject id
+        external_locations_subject_id_lut[media_url] = loc.linked_id
+        # return the mime / url combination for building the media location resources
+        { mime_type => media_url }
+      end
+      Subject.location_attributes_from_params(external_locations).map do |location_param|
```
Comment (camallen) [inline app/operations/subject_groups/create.rb:109]: this method is getting pretty big / can probably be broken up - i'll put that on the todo list (feel free to remind me in review as well)

File: spec/operations/subject_groups/create_spec.rb [hunk 2]
```diff
@@ -10,28 +10,47 @@
...
+    operation.run!(params)
   end
 
   it 'creates creates a valid subject_group' do
     expect(created_subject_group).to be_valid
   end
 
-  it 'raises with error if it can not find all the subject_ids' do
-    incorrect_subject_ids = subject_ids | ['-1']
+  it 'raises with error if it can not find all the subject_ids'  do
```
[the remark this answers is not in the document]
Comment (camallen) [reply spec/operations/subject_groups/create_spec.rb:22]:
```suggestion
  it 'raises with error if it can not find all the subject_ids' do
```

**Review by nciemniak — APPROVED** [review]
Looks good 👍 
One thing I noticed while testing on staging though - when make a GET request against `https://panoptes-staging.zooniverse.org/api/subjects/grouped?workflow_id=3412&num_rows=5&num_columns=5` (the example provided by Shaun), my request errored because it seemed to have tried to do an insert into the `subject_groups` table, despite there already being a record that matched the record that was being asked for:
```
{"errors":[{"message":"PG::UniqueViolation: 
ERROR:  duplicate key value violates unique constraint \"index_subject_groups_on_key\"
DETAIL:  Key (key)=(134598-134878-134822-134597-134787-134767-134711-134818-134662-134788-134865-134640-134623-134720-134740-134780-134717-134635-134676-134725-134873-134622-134891-134637-134817-134807-134825-134769-134712-134745-134667-134726-134649-134864-134716-134826-134816-134646-134856-134624-134747-134669-134661-134627-134881-134699-134893-134840-134685-134696-134790-134752-134603-134888-134890-134659-134751-134727-134820-134595-134700-134847-134851-134683-134858-134736-134834-134735-134762-134750-134612-134773-134703-134766-134690) already exists.: 
INSERT INTO \"subject_groups\" (\"project_id\", \"key\", \"group_subject_id\", \"created_at\", \"updated_at\") 
VALUES (1900, '134598-134878-134822-134597-134787-134767-134711-134818-134662-134788-134865-134640-134623-134720-134740-134780-134717-134635-134676-134725-134873-134622-134891-134637-134817-134807-134825-134769-134712-134745-134667-134726-134649-134864-134716-134826-134816-134646-134856-134624-134747-134669-134661-134627-134881-134699-134893-134840-134685-134696-134790-134752-134603-134888-134890-134659-134751-134727-134820-134595-134700-134847-134851-134683-134858-134736-134834-134735-134762-134750-134612-134773-134703-134766-134690', 134918, '2021-01-20 19:25:34.402096', '2021-01-20 19:25:34.402096') 
RETURNING \"id\""}]}
```

File: app/operations/subject_groups/create.rb [hunk 3]
```diff
@@ -83,12 +96,32 @@ def build_group_subject
 
     def subject_location_params
       subject_locations = find_subjects_in_order(subject_ids).map(&:locations).flatten
+      external_locations_subject_id_lut = {}
```
Comment (nciemniak) [inline app/operations/subject_groups/create.rb:99]: What does `lut` stand for in this variable name?

Comment (camallen) [reply app/operations/subject_groups/create.rb:99]: Look Up Table (LUT) - it's a pretty big variable name already but can certainly be modified if you prefer?

**Comment by camallen** [comment]
> Looks good 👍
> One thing I noticed while testing on staging though - when make a GET request against `https://panoptes-staging.zooniverse.org/api/subjects/grouped?workflow_id=3412&num_rows=5&num_columns=5` (the example provided by Shaun), my request errored because it seemed to have tried to do an insert into the `subject_groups` table, despite there already being a record that matched the record that was being asked for:
> 
> ```
> {"errors":[{"message":"PG::UniqueViolation: 
> ERROR:  duplicate key value violates unique constraint \"index_subject_groups_on_key\"
> DETAIL:  Key (key)=(134598-134878-134822-134597-134787-134767-134711-134818-134662-134788-134865-134640-134623-134720-134740-134780-134717-134635-134676-134725-134873-134622-134891-134637-134817-134807-134825-134769-134712-134745-134667-134726-134649-134864-134716-134826-134816-134646-134856-134624-134747-134669-134661-134627-134881-134699-134893-134840-134685-134696-134790-134752-134603-134888-134890-134659-134751-134727-134820-134595-134700-134847-134851-134683-134858-134736-134834-134735-134762-134750-134612-134773-134703-134766-134690) already exists.: 
> INSERT INTO \"subject_groups\" (\"project_id\", \"key\", \"group_subject_id\", \"created_at\", \"updated_at\") 
> VALUES (1900, '134598-134878-134822-134597-134787-134767-134711-134818-134662-134788-134865-134640-134623-134720-134740-134780-134717-134635-134676-134725-134873-134622-134891-134637-134817-134807-134825-134769-134712-134745-134667-134726-134649-134864-134716-134826-134816-134646-134856-134624-134747-134669-134661-134627-134881-134699-134893-134840-134685-134696-134790-134752-134603-134888-134890-134659-134751-134727-134820-134595-134700-134847-134851-134683-134858-134736-134834-134735-134762-134750-134612-134773-134703-134766-134690', 134918, '2021-01-20 19:25:34.402096', '2021-01-20 19:25:34.402096') 
> RETURNING \"id\""}]}
> ```

Yes - this was one of the bugs that is fixed by this PR (not yet on staging API), see [this diff](https://github.com/zooniverse/panoptes/compare/zooniverse:master...camallen:subject-groups-creation-bug-fix-1#diff-7673996603c9558e78cae3218bfaea6b0b968c70c97b781b2c882695a549abfe), in particular this line change
```
- subject_ids: selected_subject_ids,
+ subject_ids: per_grid_subject_ids,
```
That change ensure we pass in the individual group not the whole set of selected ids for 3 groups :facepalm:

File: app/operations/subject_groups/create.rb (hunk 3 above)
Comment (nciemniak) [reply app/operations/subject_groups/create.rb:117]: Maybe a comment next to the variable name would be helpful! e.g. `# look up table`

**Comment by nciemniak** [comment]
Aha! that makes sense - didnt notice that that was addressed. Great!

**Pull request merged**